src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs

```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+            // So to handle this, we are returning a list of port names from both locations.
+            // ToDo: If duplicates wanted to be removed, enumerate the rgistery and try and match & remove duplicate
+            //   HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Enum\USB is one location, but matching criteria not clear!
+            //   HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Control\COM Name Arbiter\Devices (empty on Win10IoT)
+
+            HashSet<string> resultPortNames = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+
+            // If running on Windows IoT, search and add the Serial Devices from QueryDosDevice
+            // If this is allowed to happen on Windows devices where port detection happens normally, duplicate Serial Port names are added.
+            // This causes issues with System.IO.Ports.Tests
+            // Port detection broken on Windows IoT (Does not initialise registry with COM port names), so use QueryDosDevice
+            if (RuntimeInformation.OSArchitecture == Architecture.Arm || RuntimeInformation.OSArchitecture == Architecture.Arm64)
+            {
+                // Query Interop QueryDosDevice()
+                //
+                // Is GUID class ID for COM ports https://docs.microsoft.com/en-us/windows-hardware/drivers/install/guid-devinterface-comport
+                // Typical DosDevice serial port name 'USB#VID_1B4F&PID_214F&MI_00#6&381731fd&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}'
+                //
+                foreach (string dosName in QueryDosDeviceComPorts(GuidDevInterfaceComPort))
+                {
+                    resultPortNames.Add(dosName);
+                }
+            }
+            else
+            {
+                // Query registry register COM[x] ports
+                //
+                using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))
+                {
+                    if (serialKey != null)
+                    {
+                        foreach (string valueName in serialKey.GetValueNames())
+                        {
+                            string portName = ((string)serialKey.GetValue(valueName));
+
+                            // Add the key's value (portname) if it is valid (Win10IoT places garbage in this Reg Key)
+                            // Filters corrupt chars for COM[x] devices in Registry on Win10IoT
+                            if (portName.StartsWith("COM", StringComparison.InvariantCultureIgnoreCase))
+                            {
+                                resultPortNames.Add(portName);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return resultPortNames.ToArray();
+        }
+
+        private static string[] QueryDosDeviceComPorts(string filterGuid)
```

Comment:
This method is **very** costly. On my current machine I've got 798 devices. We should use `ArrayPool` for this and walk through the buffer with `ReadOnlySpan`. We'll have a convenience method for this eventually (#26528), but it isn't too terribly hard to do to write a loop to slice repeatedly at `IndexOf`.

Reply:
The buffer needed to contain all of those is 54K. With the iterations here and marshaling I'd be surprised if we allocate less than a megabyte.

Reply:
Haven't mocked that mocked devices yet, but was the performance hit from Registry enumeration or HashSet?

Im questioning if we need to use HashSet here at all?
From testing, it seems that even if a duplicate COM port name exists (have to be user over ridden in devices [with ignoring warnings!]) the duplicate name does not get created in the registry (cant test Win7 yet)
Since a device name collision would be a badly configured system with un-expected results anyway, is it worth filtering for this unusual condition, or Im I missing something?
![image](https://user-images.githubusercontent.com/11331370/66246335-b9ad7d00-e752-11e9-9bdf-2f5c37a36c34.png)
![image](https://user-images.githubusercontent.com/11331370/66246348-d053d400-e752-11e9-9aa8-d66681af5076.png)
![image](https://user-images.githubusercontent.com/11331370/66246374-e2357700-e752-11e9-815e-83a6973c28e9.png)
![image](https://user-images.githubusercontent.com/11331370/66246395-ed88a280-e752-11e9-8811-a0ca1a76ce06.png)

Reply:
Im not great at this, but trying to diagnose this expensive function so I have a baseline to compare any changes to. (eg before going to try using ArrayPool etc)

On PC with >700 Registry serial port entries, appears the Interop call to Win32.Registry:Key:GetValue (reading the com port names) is taking the most?
How do to over come that? If this is where we need to read the info from?

Or, as there is more than one way to get com-port names (eg QueryDosDeviceComPorts() used to work around the garbage chars problem in Win10IoT) We just use that for all Windows targets? Even though it is still doing Interop calls, maybe its more efficient with its fill buffer in chunks/repeat approach. But I cant mock 700 real serial devices here!
![image](https://user-images.githubusercontent.com/11331370/66262997-cacdbb00-e832-11e9-842c-0c3f1c75ce7b.png)